ndcube/utils/wcs.py

```diff
@@ -197,18 +202,37 @@ def _wcs_slicer(wcs, missing_axis, item):
                 item_checked.append(slice(0, 1))
         # if all are slice in the item tuple
         if _all_slice(item_checked):
-            new_wcs = wcs.slice((item_checked))
+            item_ = (item_checked)
         # if all are not slices some of them are int then
         else:
             # this will make all the item in item_checked as slice.
             item_ = _slice_list(item_checked)
-            new_wcs = wcs.slice(item_)
             for i, it in enumerate(item_checked):
                 if isinstance(it, int):
                     missing_axis[i] = True
+    else:
+        raise TypeError("item type is {0}.  Must be int, slice, or tuple of ints and/or slices.".format(type(item)))
     # returning the reverse list of missing axis as in the item here was reverse of
     # what was inputed so we had a reverse missing_axis.
-    return new_wcs, missing_axis[::-1]
+    dropped_coords = [] # Initiating new list to collect dropped coords in the process of slicing.
+    # Checking item_ slices for dropped axes if any
+    for i, slice_element in enumerate(item_):
+        if missing_axis[i] is False and (slice_element.stop - slice_element.start) == 1:
```

Comment:
I think the solution for the ```TypeError: unsupported operand type(s) for -: 'NoneType' and 'NoneType'``` we were seeing is to break up the above ```if``` statement into two.

The first part of the statement (```missing_axis[i] is False```) checks whether the axis is ```missing```, i.e. is it dependent.  If not, we proceed.  The next part (```(slice_element.stop - slice_element.start) == 1```) is checking whether we are slicing out that dimension by slicing it with a single index.  We can tell this if ```slice_element.stop - slice_element.start == 1```.  However, in some cases, the slices can have ```None``` in them which for the ```start``` parameter represents the first element.  For the ```stop``` parameter ```None``` represents the last element and for the ```slice.step``` parameter is represents ```1```.  So when a ```None``` appears in a ```slice``` we need to convert this to the relevant number and then determine whether this represents a case where ```slice_element.stop - slice_element.start == 1```.

So, we start of which something like
```python
if missing_axis[i] is False:
    # Determine the start index
    if slice_element.start is None:
        slice_start = 0
    else:
        slice_start = slice_element.start
    # Determine the stop index.
    if slice_element.stop is None:
        slice_stop = wcs._naxis[i] # wcs._naxis is a list of the length of each axis
    else:
        slice_stop = slice_element.stop
    # Determine the slice's step.  (We will use this is a later version of this code to be more thorough.  For now we'll calculate it and not use it.)
    if slice_element.step is None:
        slice_step = 1
    else:
        slice_step = slice_element.step
    if slice_stop - slice_start == 1:
        # proceed with rest of if statement...
```


**More Advanced Info on ```step```**
A slice's ```step``` attribute gives the increment in which the slice should cover the range defined by the ```start``` and ```stop``` parameters.  So ```slice(3, 10, 3)``` means "slice out every third element between the 3rd and 10th elements".  So in code that would give:
```python
>>> l = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13] # Define a list to slice
>>> s = slice(3, 10, 3)
>>> l[s]
[3, 6, 9]
```
Here's another case where ```s.stop - s.stop != 1``` but the result of the sliced list can be a single element.
```python
>>> l = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13] # Define a list to slice
>>> s = slice(3, 5, 3) # slice every 3rd element between the 3rd and 5th, i.e. just the 3rd.
>>> l[s]
3
```
Therefore to catch these cases in the above code, change the ``` if slice_stop - slice_start == 1``` line to
```python
if (slice_stop - slice_start) / slice_step <= 1:
    # proceed with rest of if statement.
```

If you then try some of the cases for which you've been getting the ```TypeError: unsupported operand type(s) for -: 'NoneType' and 'NoneType'``` error, you should get a different result.

Reply:
Yup, have implemented the new code snippets, am testing the cases again...

Reply:
When I am testing the case `utils_wcs._wcs_slicer(w2, [False, True, False], (slice(0, 2), 1, slice(0,2)))`, instead of getting the `NoneType` error, I now get the following `IndexError`:  
```
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-28-f6be69d0cded> in <module>
----> 1 utils_wcs._wcs_slicer(w2, [False, True, False], (slice(0, 2), 1, slice(0,2)))

~/ndcube-git/ndcube/utils/wcs.py in _wcs_slicer(wcs, missing_axis, item)
    250                 # now complete.
    251                 # The dropped_coords's first variable is the IVOA axis name corresponding to the CTYPE.
--> 252     new_wcs = wcs.slice(item_)
    253     return new_wcs, missing_axis[::-1], dropped_coords
    254 

/anaconda3/envs/ndcube-dev/lib/python3.7/site-packages/astropy/wcs/wcs.py in slice(self, view, numpy_order)
   2994                 # range requires integers but the other attributes can also
   2995                 # handle arbitrary values, so this needs to be in a try/except.
-> 2996                 nitems = len(builtins.range(self._naxis[wcs_index])[iview])
   2997             except TypeError as exc:
   2998                 if 'indices must be integers' not in str(exc):

IndexError: list index out of range
```

Reply:
Also, for the command `utils_wcs._wcs_slicer(w2, [False, True, False], (None, None, None))`, I am getting the following `ValueError`: 
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-34-063e46e52a66> in <module>
----> 1 utils_wcs._wcs_slicer(w2, [False, True, False], (None, None, None))

~/ndcube-git/ndcube/utils/wcs.py in _wcs_slicer(wcs, missing_axis, item)
    240                 pix_coords[i] = slice_element.start # Enter pixel coordinate for this axis.
    241                 # Since we are only dealing with independent axes the other axes can remain at 0.
--> 242                 real_world_coords = wcs.all_pix2world(*pix_coords, 0)[i]
    243                 # Unravel the arguments in the pix_coords array using the * prefix.
    244                 # Added in index to obtain the i-th element in the resultant real world coords list of arrays.

/anaconda3/envs/ndcube-dev/lib/python3.7/site-packages/astropy/wcs/wcs.py in all_pix2world(self, *args, **kwargs)
   1295     def all_pix2world(self, *args, **kwargs):
   1296         return self._array_converter(
-> 1297             self._all_pix2world, 'output', *args, **kwargs)
   1298     all_pix2world.__doc__ = """
   1299         Transforms pixel coordinates to world coordinates.

/anaconda3/envs/ndcube-dev/lib/python3.7/site-packages/astropy/wcs/wcs.py in _array_converter(self, func, sky, ra_dec_order, *args)
   1270                     "(coords[N][{0}], origin)".format(self.naxis))
   1271             if xy.shape == () or len(xy.shape) == 1:
-> 1272                 return _return_list_of_arrays([xy], origin)
   1273             return _return_single_array(xy, origin)
   1274 

/anaconda3/envs/ndcube-dev/lib/python3.7/site-packages/astropy/wcs/wcs.py in _return_list_of_arrays(axes, origin)
   1238             if ra_dec_order and sky == 'input':
   1239                 xy = self._denormalize_sky(xy)
-> 1240             output = func(xy, origin)
   1241             if ra_dec_order and sky == 'output':
   1242                 output = self._normalize_sky(output)

ValueError: Wrong number of dimensions in input array.  Expected 2.
```
Do you think I can correct this by invoking the NDCube's [`pixel_to_world`](http://docs.sunpy.org/projects/ndcube/en/stable/api/ndcube.NDCube.html#ndcube.NDCube.pixel_to_world/) instead of astropy's [`all_pix2world`](http://docs.astropy.org/en/stable/api/astropy.wcs.WCS.html#astropy.wcs.WCS.all_pix2world)?

Reply:
Hi @kakirastern.  No, ```NDCube.pixel_to_world``` won't help in this case.  We are at a lower level than any ```NDCube``` methods.

I suspect the reason for your ```ValueError``` is that ```(None, None, None)``` is an invalid input.  Again, the code you are working is low level and so is not as robust as a user-facing API.  It assumes that the inputs have been checked and put into the correct format by the functions that call it.

As for your ```IndexError```, I suspect that is because for some reason the WCS object you created has zeros for the NAXIS values.  NAXIS denotes the length of the axes to which the WCS axis corresponds.  The values of NAXIS need to be set to a value > 1.  Then you will only be able to slice the object within the boundaries of those lengths.  Can you show me the code you are using to build your WCS object?

Reply:
Yup, the `IndexError` must be because I have not set the NAXIS values to > 1. (Sorry, it was my bad that I somehow omitted to check the values for the NAXIS.) 

Sure, the following is the code I used to construct my WCS object...
```
w2 = wcs.WCS(naxis=3)
w2.wcs.crpix = [5.995036429954E+02, 5.995573403606E+02, 6563]
w2.wcs.cdelt = np.array([ 0.00000567890, 0.00000567890, 0.0000123456])
w2.wcs.crval = [2.432585000000E+02, -5.410897222222E+0, 6300]
w2.wcs.ctype = ["HPLT", "HPLN", "WAVE"]
```

Let me change the value settings and try the cases again. Will get back to you soon.

Reply:
OK.  A couple things to note.  I think it's better practice to initialise WCS object by creating a dictionary of the parameters and values and then feeding that dictionary into the WCS constructor.  See the example in http://docs.sunpy.org/projects/ndcube/en/stable/ndcube.html#initialization
Note how the ```NAXIS1```, ```NAXIS2``` and ```NAXIS3``` parameters are set.

Be aware that if you want to mark an axis as ```missing```, the corresponding ```NAXISX``` value must be ```1```.

Finally note that, confusingly, when accessing the attributes of the fully formed WCS object, ```WCS.naxis``` gives you the number of axes, while ```WCS._naxis``` gives you the length of each axis.  Better still, use ```WCS.pixel_shape``` as it is more reliable to not change without warning in future versions.

So in summary, build your WCS object using a dictionary which defines the ```NAXISX``` values as in http://docs.sunpy.org/projects/ndcube/en/stable/ndcube.html#initialization and try again.

Reply:
Yup, done.